src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1017 @@
+    # alternative is indeed different in relation to the pair, we document the differences between
+    # the configurations in advance:
+    # ALTS(:,best_conf,pair) = the two configurations in which J-sync differs from best_conf in relation to pair
+
+    _ALTS = np.array(
+        [
+            [[1, 2, 1], [0, 2, 0], [0, 0, 1], [1, 0, 0]],
+            [[2, 3, 3], [3, 3, 2], [3, 1, 3], [2, 1, 2]],
+        ],
+        dtype=int,
+    )
+
+    def __init__(
+        self,
+        src,
+        n_rad=None,
+        n_theta=360,
+        max_shift=0.15,
+        shift_step=1,
+        epsilon=1e-2,
+        max_iters=1000,
+        seed=None,
+        mask=True,
+        S_weighting=False,
+        J_weighting=False,
+        hist_intervals=100,
+    ):
+        """
+        Initialize object for estimating 3D orientations.
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction
+        :param n_theta: The number of points in the theta direction
+        :param max_shift: Maximum range for shifts as a proportion of resolution. Default = 0.15.
+        :param shift_step: Resolution of shift estimation in pixels. Default = 1 pixel.
+        :param epsilon: Tolerance for the power method.
+        :param max_iter: Maximum iterations for the power method.
+        :param seed: Optional seed for RNG.
+        :param mask: Option to mask `src.images` with a fuzzy mask (boolean).
+            Default, `True`, applies a mask.
+        :param S_weighting: Optionally apply probabilistic weighting
+            to the `S` matrix.
+        :param J_weighting: Optionally use `J` weights instead of
+            signs when computing `signs_times_v`.
+        :param hist_intervals: Number of histogram bins used to
+            compute triangle scores when `S_weighting` enabled.
+        """
+
+        super().__init__(
+            src,
+            n_rad=n_rad,
+            n_theta=n_theta,
+            max_shift=max_shift,
+            shift_step=shift_step,
+            mask=mask,
+        )
+
+        # Generate pair mappings
+        self._pairs, self._pairs_to_linear = all_pairs(self.n_img, return_map=True)
+
+        self.epsilon = epsilon
+        self.max_iters = max_iters
+        self.seed = seed
+
+        # Sync3N specific vars
+        self.S_weighting = S_weighting
+        self.J_weighting = J_weighting
+        self._D_null = 1e-13
+        self.hist_intervals = int(hist_intervals)
+        # Warn if histogram may be too sparse for curve fitting
+        if self.S_weighting and (src.n < hist_intervals):
+            logger.warning(
+                f"`hist_intervals` {hist_intervals} > src.n {src.n}."
+                "  Consider reducing if curve fitting is infeasable."
+            )
+
+        # Auto configure GPU
+        self.__gpu_module = None
```

Comment:
Should we have a way to force CPU even if GPU is present (maybe the user has a faulty device or something, although one would expect `import cupy` to fail there)?

Reply:
Uninstalling `cupy` would do it, but that is a little roundabout. Added a `disable_gpu` flag.  

Keep in mind this algorithm is practically unusable beyond tiny problem on the CPU.

Reply:
> Uninstalling `cupy` would do it, but that is a little roundabout. Added a `disable_gpu` flag.

Ok sounds good.

> Keep in mind this algorithm is practically unusable beyond tiny problem on the CPU.

Yeah I know, but it's just good to have the option.